gofmt failure, with Go 1.16 (Aarch64) + patching, on behalf of go-reflect package

[rook_go-reflect.diff.zip](https://github.com/rook/rook/files/6098117/rook_go-reflect.diff.zip)

Hi.
I am using Go 1.16 on an Aarch64 server.

> oceanfish81@ams1-c2:~/rook$ make    
> === helm package rook-ceph
> ==> Linting /home/oceanfish81/rook/_output/rook-ceph
> 1 chart(s) linted, 0 chart(s) failed
> Successfully packaged chart and saved it to: /home/oceanfish81/rook/_output/charts/rook-ceph-v1.5.0-alpha.0.590.g2a192300e-dirty.tgz
> === helm index
> === ensuring modules are tidied
> === go vet
> diff -u pkg/daemon/ceph/client/pool_test.go.orig pkg/daemon/ceph/client/pool_test.go
> --- pkg/daemon/ceph/client/pool_test.go.orig	2021-03-07 22:50:25.081808943 +0000
> +++ pkg/daemon/ceph/client/pool_test.go	2021-03-07 22:50:25.081808943 +0000
> @@ -16,8 +16,8 @@
>  package client
>  import (
> -	"os/exec"
>  	reflect "github.com/goccy/go-reflect"
> +	"os/exec"
>  	"strconv"
>  	"testing"
> diff -u pkg/operator/cassandra/sidecar/sidecar.go.orig pkg/operator/cassandra/sidecar/sidecar.go
> --- pkg/operator/cassandra/sidecar/sidecar.go.orig	2021-03-07 22:50:25.261806937 +0000
> +++ pkg/operator/cassandra/sidecar/sidecar.go	2021-03-07 22:50:25.261806937 +0000
> @@ -19,10 +19,10 @@
>  import (
>  	"context"
>  	"fmt"
> +	reflect "github.com/goccy/go-reflect"
>  	"net/url"
>  	"os"
>  	"os/exec"
> -	reflect "github.com/goccy/go-reflect"
>  	"time"
>  	"github.com/coreos/pkg/capnslog"
> diff -u pkg/operator/ceph/cluster/mon/health_test.go.orig pkg/operator/ceph/cluster/mon/health_test.go
> --- pkg/operator/ceph/cluster/mon/health_test.go.orig	2021-03-07 22:50:25.401805376 +0000
> +++ pkg/operator/ceph/cluster/mon/health_test.go	2021-03-07 22:50:25.401805376 +0000
> @@ -19,9 +19,9 @@
>  import (
>  	"context"
>  	"fmt"
> +	reflect "github.com/goccy/go-reflect"
>  	"io/ioutil"
>  	"os"
> -	reflect "github.com/goccy/go-reflect"
>  	"sync"
>  	"testing"
>  	"time"
> diff -u pkg/operator/ceph/cluster/mon/mon_test.go.orig pkg/operator/ceph/cluster/mon/mon_test.go
> --- pkg/operator/ceph/cluster/mon/mon_test.go.orig	2021-03-07 22:50:25.445804886 +0000
> +++ pkg/operator/ceph/cluster/mon/mon_test.go	2021-03-07 22:50:25.445804886 +0000
> @@ -19,10 +19,10 @@
>  import (
>  	"context"
>  	"fmt"
> +	reflect "github.com/goccy/go-reflect"
>  	"io/ioutil"
>  	"os"
>  	"path"
> -	reflect "github.com/goccy/go-reflect"
>  	"strconv"
>  	"strings"
>  	"sync"
> diff -u pkg/operator/ceph/cluster/operator_watchers.go.orig pkg/operator/ceph/cluster/operator_watchers.go
> --- pkg/operator/ceph/cluster/operator_watchers.go.orig	2021-03-07 22:50:25.477804529 +0000
> +++ pkg/operator/ceph/cluster/operator_watchers.go	2021-03-07 22:50:25.477804529 +0000
> @@ -18,8 +18,8 @@
>  package cluster
>  import (
> -	"os"
>  	reflect "github.com/goccy/go-reflect"
> +	"os"
>  	opcontroller "github.com/rook/rook/pkg/operator/ceph/controller"
>  	"github.com/rook/rook/pkg/operator/k8sutil"
> diff -u pkg/operator/ceph/file/filesystem_test.go.orig pkg/operator/ceph/file/filesystem_test.go
> --- pkg/operator/ceph/file/filesystem_test.go.orig	2021-03-07 22:50:25.797800962 +0000
> +++ pkg/operator/ceph/file/filesystem_test.go	2021-03-07 22:50:25.797800962 +0000
> @@ -20,10 +20,10 @@
>  	"context"
>  	"errors"
>  	"fmt"
> +	reflect "github.com/goccy/go-reflect"
>  	"io/ioutil"
>  	"os"
>  	"path"
> -	reflect "github.com/goccy/go-reflect"
>  	"strings"
>  	"testing"
> diff -u pkg/operator/nfs/controller_test.go.orig pkg/operator/nfs/controller_test.go
> --- pkg/operator/nfs/controller_test.go.orig	2021-03-07 22:50:26.173796770 +0000
> +++ pkg/operator/nfs/controller_test.go	2021-03-07 22:50:26.173796770 +0000
> @@ -18,8 +18,8 @@
>  import (
>  	"context"
> -	"path"
>  	reflect "github.com/goccy/go-reflect"
> +	"path"
>  	"testing"
>  	nfsv1alpha1 "github.com/rook/rook/pkg/apis/nfs.rook.io/v1alpha1"
> diff -u pkg/operator/nfs/provisioner_test.go.orig pkg/operator/nfs/provisioner_test.go
> --- pkg/operator/nfs/provisioner_test.go.orig	2021-03-07 22:50:26.201796458 +0000
> +++ pkg/operator/nfs/provisioner_test.go	2021-03-07 22:50:26.201796458 +0000
> @@ -18,8 +18,8 @@
>  import (
>  	"context"
> -	"os"
>  	reflect "github.com/goccy/go-reflect"
> +	"os"
>  	"testing"
>  	rookclient "github.com/rook/rook/pkg/client/clientset/versioned"
> make[1]: *** [build/makelib/golang.mk:160: go.fmt] Error 1
> make: *** [Makefile:100: build.common] Error 2

Aha! I think I found something:

> oceanfish81@ams1-c2:~/rook$ make
> === helm package rook-ceph
> ==> Linting /home/oceanfish81/rook/_output/rook-ceph
> 
> 1 chart(s) linted, 0 chart(s) failed
> Successfully packaged chart and saved it to: /home/oceanfish81/rook/_output/charts/rook-ceph-v1.5.0-alpha.0.623.gc0eacc89d-dirty.tgz
> === helm index
> === ensuring modules are tidied
> === go vet
> diff -u pkg/daemon/ceph/client/pool_test.go.orig pkg/daemon/ceph/client/pool_test.go
> --- pkg/daemon/ceph/client/pool_test.go.orig	2021-03-10 23:49:11.399329184 +0000
> +++ pkg/daemon/ceph/client/pool_test.go	2021-03-10 23:49:11.399329184 +0000
> @@ -16,8 +16,8 @@
>  package client
>  
>  import (
> -	"os/exec"
>  	reflect "github.com/goccy/go-reflect"
> +	"os/exec"
>  	"strconv"
>  	"testing"
>  
> make[1]: *** [build/makelib/golang.mk:160: go.fmt] Error 1
> make: *** [Makefile:100: build.common] Error 2
> 

[transactional2_diff4rook.log](https://github.com/rook/rook/files/6119492/transactional2_diff4rook.log)

> Hy @advancedwebdeveloper , I faced the same bug before while building the `Ceph` image.
> 
> The issue is caused because of not **proper formatting** up the Go code while doing the changes.
> Try building image with **no changes** then you will have **no errors**.
> 
> If your adding changes to it try to format your Go file correctly by saving file with `**Ctrl+ s**` in vs code.
> Looking at your errors:
> ex:`diff -u pkg/daemon/ceph/client/pool_test.go.orig pkg/daemon/ceph/client/pool_test.go`
> So inside **pool_test.go**
> 
> Try importing like this:
> 
> ```
> import (
> 	"os/exec"
> 	"strconv"
> 	"testing"
> 
> 	reflect "github.com/goccy/go-reflect"
> 
> 	"github.com/pkg/errors"
> 	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
> 	rookv1 "github.com/rook/rook/pkg/apis/rook.io/v1"
> 	exectest "github.com/rook/rook/pkg/util/exec/test"
> 	"github.com/stretchr/testify/assert"
> 
> 	"github.com/rook/rook/pkg/clusterd"
> )
> ```
> 
> Hope this solves your issue.

It did not help:

> === go vet
> diff -u pkg/daemon/ceph/client/pool_test.go.orig pkg/daemon/ceph/client/pool_test.go
> --- pkg/daemon/ceph/client/pool_test.go.orig    2021-03-12 10:38:21.223467677 +0000
> +++ pkg/daemon/ceph/client/pool_test.go 2021-03-12 10:38:21.223467677 +0000
> @@ -16,19 +16,19 @@
>  package client
> 
>  import (
> -    "os/exec"
> -    "strconv"
> -    "testing"
> +       "os/exec"
> +       "strconv"
> +       "testing"
> 
> -    reflect "github.com/goccy/go-reflect"
> +       reflect "github.com/goccy/go-reflect"
> 
> -    "github.com/pkg/errors"
> -    cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
> -    rookv1 "github.com/rook/rook/pkg/apis/rook.io/v1"
> -    exectest "github.com/rook/rook/pkg/util/exec/test"
> -    "github.com/stretchr/testify/assert"
> +       "github.com/pkg/errors"
> +       cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
> +       rookv1 "github.com/rook/rook/pkg/apis/rook.io/v1"
> +       exectest "github.com/rook/rook/pkg/util/exec/test"
> +       "github.com/stretchr/testify/assert"
> 
> -    "github.com/rook/rook/pkg/clusterd"
> +       "github.com/rook/rook/pkg/clusterd"
>  )
> 
>  func TestCreateECPoolWithOverwrites(t *testing.T) {
> make[1]: *** [build/makelib/golang.mk:160: go.fmt] Error 1
> make: *** [Makefile:100: build.common] Error 2

@ianlancetaylor, perhaps a patch for gofmt would help?

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -14,7 +14,7 @@ jobs:
 
     - uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: build rook
       run: |

--- a/.github/workflows/canary-integration-test.yml
+++ b/.github/workflows/canary-integration-test.yml
@@ -20,7 +20,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1
@@ -128,7 +128,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: install deps
       run: |
@@ -228,7 +228,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: install deps
       run: |
@@ -320,7 +320,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: install deps
       run: |
@@ -414,7 +414,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1
@@ -502,7 +502,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: install deps
       run: |
@@ -594,7 +594,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: install deps
       run: |
@@ -688,7 +688,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1
@@ -789,7 +789,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: install deps
       run: |

--- a/.github/workflows/codegen.yml
+++ b/.github/workflows/codegen.yml
@@ -10,7 +10,7 @@ jobs:
     steps:
     - uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: checkout
       uses: actions/checkout@v2

--- a/.github/workflows/crds-gen.yml
+++ b/.github/workflows/crds-gen.yml
@@ -8,7 +8,7 @@ jobs:
     steps:
     - uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: checkout
       uses: actions/checkout@v2

--- a/.github/workflows/integration-test-mgr-suite.yaml
+++ b/.github/workflows/integration-test-mgr-suite.yaml
@@ -15,7 +15,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1

--- a/.github/workflows/integration-test-multi-cluster-suite.yaml
+++ b/.github/workflows/integration-test-multi-cluster-suite.yaml
@@ -23,7 +23,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1
@@ -89,7 +89,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1

--- a/.github/workflows/integration-test-smoke-suite.yaml
+++ b/.github/workflows/integration-test-smoke-suite.yaml
@@ -23,7 +23,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1
@@ -88,7 +88,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1

--- a/.github/workflows/integration-test-upgrade-suite.yaml
+++ b/.github/workflows/integration-test-upgrade-suite.yaml
@@ -23,7 +23,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1
@@ -88,7 +88,7 @@ jobs:
     - name: setup golang
       uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: setup minikube
       uses: manusa/actions-setup-minikube@v2.3.1

--- a/.github/workflows/mod-check.yml
+++ b/.github/workflows/mod-check.yml
@@ -13,11 +13,10 @@ jobs:
 
     - uses: actions/setup-go@v2
       with:
-        go-version: 1.15
+        go-version: 1.16
 
     - name: run mod check
       run: GOPATH=$(go env GOPATH) make -j $(nproc) mod.check
 
     - name: validate modcheck
       run: tests/scripts/validate_modified_files.sh modcheck
-

--- a/build/makelib/golang.mk
+++ b/build/makelib/golang.mk
@@ -48,7 +48,7 @@ GO_TEST_FLAGS ?=
 # ====================================================================================
 # Setup go environment
 
-GO_SUPPORTED_VERSIONS ?= 1.15
+GO_SUPPORTED_VERSIONS ?= 1.15|1.16
 
 GO_PACKAGES := $(foreach t,$(GO_SUBDIRS),$(GO_PROJECT)/$(t)/...)
 GO_INTEGRATION_TEST_PACKAGES := $(foreach t,$(GO_INTEGRATION_TESTS_SUBDIRS),$(GO_PROJECT)/$(t)/integration)
@@ -91,8 +91,6 @@ endif
 # NOTE: the install suffixes are matched with the build container to speed up the
 # the build. Please keep them in sync.
 
-# we run go build with -i which on most system's would want to install packages
-# into the system's root dir. using our own pkg dir avoid this
 ifneq ($(GO_PKG_DIR),)
 GO_PKG_BASE_DIR := $(abspath $(GO_PKG_DIR)/$(PLATFORM))
 GO_PKG_STATIC_FLAGS := -pkgdir $(GO_PKG_BASE_DIR)_static
@@ -121,8 +119,8 @@ go.init:
 .PHONY: go.build
 go.build:
 	@echo === go build $(PLATFORM)
-	$(foreach p,$(GO_STATIC_PACKAGES),@CGO_ENABLED=0 $(GO) build -v -i -o $(GO_OUT_DIR)/$(lastword $(subst /, ,$(p)))$(GO_OUT_EXT) $(GO_STATIC_FLAGS) $(p)${\n})
-	$(foreach p,$(GO_TEST_PACKAGES),@CGO_ENABLED=0 $(GO) test -v -i -c -o $(GO_TEST_OUTPUT)/$(lastword $(subst /, ,$(p)))$(GO_OUT_EXT) $(GO_STATIC_FLAGS) $(p)${\n})
+	$(foreach p,$(GO_STATIC_PACKAGES),@CGO_ENABLED=0 $(GO) build -v -o $(GO_OUT_DIR)/$(lastword $(subst /, ,$(p)))$(GO_OUT_EXT) $(GO_STATIC_FLAGS) $(p)${\n})
+	$(foreach p,$(GO_TEST_PACKAGES),@CGO_ENABLED=0 $(GO) test -v -c -o $(GO_TEST_OUTPUT)/$(lastword $(subst /, ,$(p)))$(GO_OUT_EXT) $(GO_STATIC_FLAGS) $(p)${\n})
 
 .PHONY: go.install
 go.install:
@@ -133,7 +131,7 @@ go.install:
 go.test.unit: $(GOJUNIT)
 	@echo === go test unit-tests
 	@mkdir -p $(GO_TEST_OUTPUT)
-	CGO_ENABLED=0 $(GOHOST) test -v -i -cover $(GO_STATIC_FLAGS) $(GO_PACKAGES)
+	CGO_ENABLED=0 $(GOHOST) test -v -cover $(GO_STATIC_FLAGS) $(GO_PACKAGES)
 	CGO_ENABLED=0 $(GOHOST) test -v -cover $(GO_TEST_FLAGS) $(GO_STATIC_FLAGS) $(GO_PACKAGES) 2>&1 | tee $(GO_TEST_OUTPUT)/unit-tests.log
 	@cat $(GO_TEST_OUTPUT)/unit-tests.log | $(GOJUNIT) -set-exit-code > $(GO_TEST_OUTPUT)/unit-tests.xml
 

--- a/images/cross/Dockerfile
+++ b/images/cross/Dockerfile
@@ -37,8 +37,8 @@ RUN DEBIAN_FRONTEND=noninteractive apt-get update && \
     DEBIAN_FRONTEND=noninteractive apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
 
 # install golang from the official repo
-RUN GO_VERSION=1.15.7 && \
-    GO_HASH=0d142143794721bb63ce6c8a6180c4062bcf8ef4715e7d6d6609f3a8282629b3 && \
+RUN GO_VERSION=1.16.3 && \
+    GO_HASH=951a3c7c6ce4e56ad883f97d9db74d3d6d80d5fec77455c6ada6c1f7ac4776d2 && \
     curl -fsSL https://dl.google.com/go/go${GO_VERSION}.linux-amd64.tar.gz -o golang.tar.gz && \
     echo "${GO_HASH}  golang.tar.gz" | sha256sum -c - && \
     tar -C /usr/local -xzf golang.tar.gz && \

